File: DotNet/Messaging/Storage/NearestServerProvider.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace Bollywell.Hydra.Messaging.Storage
{
    /// <summary>
    /// Use the Hydra server with the smallest response time. Switch if another one is more than Tolerance milliseconds faster.
    /// </summary>
    public class NearestServerProvider : PollingProviderBase
    {

        /// <summary>
        /// Only change servers if the distance is more than Tolerance milliseconds less than the distance to the current server.
        /// </summary>
        public long Tolerance = 50;

        /// <summary>
        /// Initialise messaging. Must be called before any attempt to send or listen.
        /// </summary>
        /// <param name="hydraServer">Hydra server to communicate with</param>
        /// <param name="database">Name of the messaging database. Defaults to "hydra"</param>
        /// <param name="port">Port number of the messaging database. Defaults to 5984</param>
        public NearestServerProvider(string hydraServer, string database = null, int? port = null) 
            : this(new List<string> {hydraServer}, database, port) {}

        /// <summary>
        /// Initialise messaging. Must be called before any attempt to send or listen.
        /// </summary>
        /// <param name="hydraServers">Hydra servers to communicate with</param>
        /// <param name="database">Name of the messaging database. Defaults to "hydra"</param>
        /// <param name="port">Port number of the messaging database. Defaults to 5984</param>
        public NearestServerProvider(IEnumerable<string> hydraServers, string database = null, int? port = null)
            : this(hydraServers.Select(s => new CouchDbStore(s, database, port))) {}

        /// <summary>
        /// Initialise messaging. Must be called before any attempt to send or listen.
        /// </summary>
        /// <param name="stores">Hydra stores to communicate with</param>
        public NearestServerProvider(IEnumerable<IStore> stores)
            : base(stores) {}

        protected override void OnDistanceInfo(ServerDistanceInfo sdi)
        {
            if (sdi.IsReachable && (HydraServer == null || (HydraServer != sdi.Name && sdi.Distance < Distances.ServerInfo[HydraServer].Distance - Tolerance))) {
                // There is a better server than the current one
                HydraServer = sdi.Name;
            } else if (!sdi.IsReachable && HydraServer == sdi.Name) {
                // The current server is no longer responding - replace with the nearest reachable one, or null if none are reachable
                HydraServer = Distances.ServerInfo.Values.Where(sdi1 => sdi1.IsReachable).OrderBy(sdi1 => sdi1.Distance).Select(sdi1 => sdi1.Name).FirstOrDefault();
            } else if (!Initialised && !sdi.IsReachable && Distances.ServerInfo.Count == StoreDict.Count) {
                // All servers are unreachable so initialisation is done
                FinishedInitialisation();
            }
        }

        #region Measure distance

        protected override void InitDistance(IEnumerable<string> servers)
        {
            // Do a one-off poll of everything and discard the results, but only wait 1.5 seconds in case of very slow responses. For some reason the first connection to
            // a CouchDb server on localhost takes one second for the TCP connect phase, so this gets over that initial slow poll.
            // TODO: convert to Rx - see http://cm-bloggers.blogspot.co.uk/2011/02/reactive-extensions.html
            Task.WaitAll(servers.Select(server => Task.Factory.StartNew(() => { MeasureDistance(server); })).ToArray(), 1500);
        }

        #endregion

    }
}

File: Java/hydra/src/main/java/uk/co/shastra/hydra/messaging/storage/NearestServerProvider.java
package uk.co.shastra.hydra.messaging.storage;

import rx.Observable;
import rx.util.functions.Action1;

public class NearestServerProvider extends PollingProviderBase {

	/**
	 * Measure distance to server on a separate thread
	 *
	 */
	private class InitServer extends Thread {
		private String server;

		public InitServer(String server) {
			this.server = server;
		}

		@Override
		public void run() {
			getMeasureDistance().call(server);
		}
		
	}
	
    public long tolerance = 50;

    public NearestServerProvider(String hydraServer) { super(hydraServer); }
    public NearestServerProvider(String hydraServer, String database) { super(hydraServer, database); }
    public NearestServerProvider(String hydraServer, String database, Integer port) { super(hydraServer, database, port); }

    public NearestServerProvider(Iterable<String> hydraServers, String database) { super(hydraServers, database); }
    public NearestServerProvider(Iterable<String> hydraServers, String database, Integer port) { super(hydraServers, database, port); }

    public NearestServerProvider(Iterable<Store> stores) { super(stores); }

    
	@Override
	protected void doSubscription(Observable<ServerDistanceInfo> observable) {
		observable.subscribe(new Action1<ServerDistanceInfo>() {
			@Override public void call(ServerDistanceInfo sdi) {
		        if (sdi.isReachable && (getHydraServer() == null || 
		        		(!getHydraServer().equals(sdi.name) && sdi.distance < distances.getServerInfo().get(getHydraServer()).distance - tolerance))) {
		            // There is a better server than the current one
		            setHydraServer(sdi.name);
		        } else if (!sdi.isReachable && getHydraServer().equals(sdi.name)) {
		            // The current server is no longer responding - replace with the nearest reachable one, or null if none are reachable
		        	ServerDistanceInfo nearest = null;
		        	for (ServerDistanceInfo candidate : distances.getServerInfo().values()) {
		        		if (candidate.isReachable && (nearest == null || candidate.distance < nearest.distance))
		        			nearest = candidate;
		        	}
		        	setHydraServer(nearest == null ? null : nearest.name);
		        } else if (!initialised && !sdi.isReachable && distances.getServerInfo().size() == storeDict.size()) {
		            // All servers are unreachable so initialisation is done
		            finishedInitialisation();
		        }
			}
		});
	}

	@Override
	protected Action1<Iterable<String>> getInitDistance() {
        // Do a one-off poll of everything and discard the results, but only wait 1.5 seconds in case of very slow responses. For some reason the first connection to
        // a CouchDb server on localhost takes one second for the TCP connect phase, so this gets over that initial slow poll.
		return new Action1<Iterable<String>>() {
			@Override
			public void call(Iterable<String> servers) {
				for (String server : servers) {
					Thread thread = new InitServer(server);
					thread.start();
					try {
						thread.join(1500);
					} catch (InterruptedException e) {}
				}
			}
		};
	}

}
